File: api/validation.py
"""Server-side field validation per LEAD_API_CONTRACT.md sections 3, 4, and 8.

`validate_lead` returns a (normalized_fields, field_errors) pair. When
field_errors is empty, normalized_fields carries every request field ready
to store; the caller adds the server-generated fields (submitted_at,
privacy_notice_version, status).
"""

from __future__ import annotations

import re

NAME_MIN_LENGTH = 3
NAME_MAX_LENGTH = 100
NAME_ALLOWED_EXTRA_CHARS = " '’-."  # space, ' and curly apostrophe, hyphen, period

EMAIL_MAX_LENGTH = 254
EMAIL_LOCAL_MAX_LENGTH = 64
EMAIL_PATTERN = re.compile(
    r"^[a-z0-9._%+-]+@[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?"
    r"(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)*\.[a-z]{2,63}$"
)

PHONE_STRIP_PATTERN = re.compile(r"[ \-().]")
PHONE_PATTERN = re.compile(r"^\+[1-9][0-9]{7,14}$")

SCREENING_ANSWER_CODES = frozenset(
    {
        "ready_to_invest",
        "needs_investment_info",
        "no_capacity_now",
        "exploring",
    }
)
LANGUAGE_CODES = frozenset({"es", "en"})


def validate_lead(payload: dict) -> tuple[dict, dict]:
    normalized: dict = {}
    errors: dict = {}

    _validate_name(payload, normalized, errors)
    _validate_email(payload, normalized, errors)
    _validate_phone(payload, normalized, errors)
    _validate_choice(payload, normalized, errors, "screening_answer", SCREENING_ANSWER_CODES)
    _validate_choice(payload, normalized, errors, "language", LANGUAGE_CODES)
    _validate_consent(payload, normalized, errors)

    return normalized, errors


def _get_required_string(payload: dict, field: str, errors: dict) -> str | None:
    value = payload.get(field)
    if value is None:
        errors[field] = "required"
        return None
    if not isinstance(value, str):
        errors[field] = "invalid_type"
        return None
    if value.strip() == "":
        errors[field] = "required"
        return None
    return value


def _validate_name(payload: dict, normalized: dict, errors: dict) -> None:
    value = _get_required_string(payload, "name", errors)
    if value is None:
        return

    collapsed = re.sub(r"\s+", " ", value.strip())

    if len(collapsed) < NAME_MIN_LENGTH:
        errors["name"] = "too_short"
        return
    if len(collapsed) > NAME_MAX_LENGTH:
        errors["name"] = "too_long"
        return
    if not all(ch.isalpha() or ch in NAME_ALLOWED_EXTRA_CHARS for ch in collapsed):
        errors["name"] = "invalid_characters"
        return

    words = collapsed.split(" ")
    if len(words) < 2 or not all(word[:1].isalpha() for word in words):
        errors["name"] = "full_name_required"
        return

    normalized["name"] = collapsed


def _validate_email(payload: dict, normalized: dict, errors: dict) -> None:
    value = _get_required_string(payload, "email", errors)
    if value is None:
        return

    candidate = value.strip().lower()
    local_part = candidate.split("@", 1)[0]

    if len(candidate) > EMAIL_MAX_LENGTH or len(local_part) > EMAIL_LOCAL_MAX_LENGTH:
        errors["email"] = "too_long"
        return
    if not EMAIL_PATTERN.match(candidate):
        errors["email"] = "invalid_format"
        return
    if local_part.startswith(".") or local_part.endswith(".") or ".." in local_part:
        errors["email"] = "invalid_format"
        return

    normalized["email"] = candidate


def _validate_phone(payload: dict, normalized: dict, errors: dict) -> None:
    value = _get_required_string(payload, "phone", errors)
    if value is None:
        return

    candidate = PHONE_STRIP_PATTERN.sub("", value)
    if not PHONE_PATTERN.match(candidate):
        errors["phone"] = "invalid_format"
        return

    normalized["phone"] = candidate


def _validate_choice(
    payload: dict, normalized: dict, errors: dict, field: str, allowed_codes: frozenset
) -> None:
    value = _get_required_string(payload, field, errors)
    if value is None:
        return
    if value not in allowed_codes:
        errors[field] = "invalid_option"
        return
    normalized[field] = value


def _validate_consent(payload: dict, normalized: dict, errors: dict) -> None:
    if "consent" not in payload or payload["consent"] is None:
        errors["consent"] = "required"
        return

    value = payload["consent"]
    if not isinstance(value, bool):
        errors["consent"] = "invalid_type"
        return
    if value is not True:
        errors["consent"] = "consent_required"
        return

    normalized["consent"] = True

File: src/lib/lead/validation.ts
// Client-side validation implementing every rule in
// docs/specs/LEAD_API_CONTRACT.md section 3 (normalization included), plus
// the screening_answer/language/consent rules from sections 4 and 8.
//
// Mirrors `api/validation.py` field-for-field so client and server reject
// and normalize the same inputs (LEAD_API_CONTRACT.md is the source of
// truth; `api/validation.py` is only a cross-check). If the two ever
// disagree, that is a contract violation to raise, not a difference to copy.
import { isLanguage, type Language } from '../../i18n';
import {
  isScreeningAnswerCode,
  type FieldErrorCode,
  type LeadFieldErrors,
  type LeadRequestPayload,
  type ScreeningAnswerCode,
} from './types';

export const NAME_MIN_LENGTH = 3;
export const NAME_MAX_LENGTH = 100;
// space, straight and curly apostrophe, hyphen, period
const NAME_ALLOWED_EXTRA_CHARS = new Set([' ', "'", '’', '-', '.']);

export const EMAIL_MAX_LENGTH = 254;
export const EMAIL_LOCAL_MAX_LENGTH = 64;
const EMAIL_PATTERN =
  /^[a-z0-9._%+-]+@[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)*\.[a-z]{2,63}$/;

const PHONE_STRIP_PATTERN = /[ \-().]/g;
const PHONE_PATTERN = /^\+[1-9][0-9]{7,14}$/;

const LETTER_PATTERN = /\p{L}/u;

export interface FieldResult<T> {
  value?: T;
  error?: FieldErrorCode;
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === 'string';
}

function requiredString(value: unknown): FieldResult<string> {
  if (value === null || value === undefined) {
    return { error: 'required' };
  }
  if (!isNonEmptyString(value)) {
    return { error: 'invalid_type' };
  }
  if (value.trim() === '') {
    return { error: 'required' };
  }
  return { value };
}

export function validateName(raw: unknown): FieldResult<string> {
  const required = requiredString(raw);
  if (required.error || required.value === undefined) {
    return required;
  }

  const collapsed = required.value.trim().replace(/\s+/g, ' ');

  if (collapsed.length < NAME_MIN_LENGTH) {
    return { error: 'too_short' };
  }
  if (collapsed.length > NAME_MAX_LENGTH) {
    return { error: 'too_long' };
  }
  for (const ch of collapsed) {
    if (!LETTER_PATTERN.test(ch) && !NAME_ALLOWED_EXTRA_CHARS.has(ch)) {
      return { error: 'invalid_characters' };
    }
  }

  const words = collapsed.split(' ');
  if (words.length < 2 || !words.every((word) => LETTER_PATTERN.test(word.charAt(0)))) {
    return { error: 'full_name_required' };
  }

  return { value: collapsed };
}

export function validateEmail(raw: unknown): FieldResult<string> {
  const required = requiredString(raw);
  if (required.error || required.value === undefined) {
    return required;
  }

  const candidate = required.value.trim().toLowerCase();
  const localPart = candidate.split('@', 1)[0];

  if (candidate.length > EMAIL_MAX_LENGTH || localPart.length > EMAIL_LOCAL_MAX_LENGTH) {
    return { error: 'too_long' };
  }
  if (!EMAIL_PATTERN.test(candidate)) {
    return { error: 'invalid_format' };
  }
  if (localPart.startsWith('.') || localPart.endsWith('.') || localPart.includes('..')) {
    return { error: 'invalid_format' };
  }

  return { value: candidate };
}

export function validatePhone(raw: unknown): FieldResult<string> {
  const required = requiredString(raw);
  if (required.error || required.value === undefined) {
    return required;
  }

  const candidate = required.value.replace(PHONE_STRIP_PATTERN, '');
  if (!PHONE_PATTERN.test(candidate)) {
    return { error: 'invalid_format' };
  }

  return { value: candidate };
}

export function validateScreeningAnswer(raw: unknown): FieldResult<ScreeningAnswerCode> {
  const required = requiredString(raw);
  if (required.error || required.value === undefined) {
    return { error: required.error };
  }
  if (!isScreeningAnswerCode(required.value)) {
    return { error: 'invalid_option' };
  }
  return { value: required.value };
}

export function validateLanguage(raw: unknown): FieldResult<Language> {
  const required = requiredString(raw);
  if (required.error || required.value === undefined) {
    return { error: required.error };
  }
  if (!isLanguage(required.value)) {
    return { error: 'invalid_option' };
  }
  return { value: required.value };
}

export function validateConsent(raw: unknown): FieldResult<true> {
  if (raw === null || raw === undefined) {
    return { error: 'required' };
  }
  if (typeof raw !== 'boolean') {
    return { error: 'invalid_type' };
  }
  if (raw !== true) {
    return { error: 'consent_required' };
  }
  return { value: true };
}

export interface RawLeadFormInput {
  name: unknown;
  email: unknown;
  phone: unknown;
  screening_answer: unknown;
  language: unknown;
  consent: unknown;
}

export interface LeadFormValidationResult {
  fieldErrors: LeadFieldErrors;
  values: Partial<LeadRequestPayload>;
}

/** Validates every field per contract section 3/4/8. Used for the full submit payload. */
export function validateLeadForm(input: RawLeadFormInput): LeadFormValidationResult {
  const fieldErrors: LeadFieldErrors = {};
  const values: Partial<LeadRequestPayload> = {};

  const name = validateName(input.name);
  if (name.error) fieldErrors.name = name.error;
  else values.name = name.value;

  const email = validateEmail(input.email);
  if (email.error) fieldErrors.email = email.error;
  else values.email = email.value;

  const phone = validatePhone(input.phone);
  if (phone.error) fieldErrors.phone = phone.error;
  else values.phone = phone.value;

  const screeningAnswer = validateScreeningAnswer(input.screening_answer);
  if (screeningAnswer.error) fieldErrors.screening_answer = screeningAnswer.error;
  else values.screening_answer = screeningAnswer.value;

  const language = validateLanguage(input.language);
  if (language.error) fieldErrors.language = language.error;
  else values.language = language.value;

  const consent = validateConsent(input.consent);
  if (consent.error) fieldErrors.consent = consent.error;
  else values.consent = consent.value;

  return { fieldErrors, values };
}
